gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/OverflowOplog.java

```diff
@@ -1409,9 +1407,16 @@ public void clear() {
       this.value = null;
     }
 
-    private void write(ByteBuffer buffer, final int byteLength) throws IOException {
-      final int position = buffer.position();
+    private void write(ByteBuffer buffer) throws IOException {
       ByteBuffer bb = getOLF().writeBuf;
+      if (!this.value.buffer.writeSerializationHeader(buffer, bb)) {
+        flush();
+        if (!this.value.buffer.writeSerializationHeader(buffer, bb)) {
```

Comment:
At the GemFire layer Oplog normally expects either: a) a serialized object in which case EntryBits for the entry will have the SERIALIZED flag, b) a raw byte[] value

Original code flow was like this:

a) Oplog would serialize the object into a temporary DataOutput, then create a flat byte[] out of it with corresponding EntryBits (2 copies)

b) Then write the byte[] to the channel (another copy to native kernel file buffer which is then written to file)

Likewise Oplog read would first read a serialized byte[] from file (one copy), then deserialize the object from a DataInput on top of this byte[] (second copy).

As part of SNAP-1427 had made two changes:

a) Introduced the SerializedDiskBuffer class which basically says that "I know how to write my serialized version to Oplog file channel directly". It has a "write(OutputStreamChannel)" method which implementations can use. See BlobHelper.serializeToBuffer on how serialization is handled now.

b) Replaced Oplog writer/reader with ChannelBufferUnsafeOutputStream (with slight extension class FileChannelOutputStream to track number of written bytes). The base OutputStreamChannel class is an extension to OutputStream which also provides a "write(ByteBuffer)" method like in FileChannel nio API. So it can directly write ByteBuffers to the FileChannel and the ChannelBufferUnsafeOutputStream uses direct buffers for buffering writes in serialization that avoids a copy. Apart from this, the implementation also has the intelligence to directly write large DirectByteBuffers to the channel without any buffering -- see OutputStreamChannel.writeBuffered.

So with these two it was able to avoid all the 3 extra copies in writing like done in ColumnFormatValue.write(OutputStreamChannel) method which will end up streaming the in-memory buffer directly to the file especially for off-heap (on-heap ByteBuffer streaming will still create one copy at kernel level).

For reads a ByteBufferDataInput is used which reads the serialized object into a direct ByteBuffer and provides a DataInput on top of it (see BlobHelper.deserializeBuffer). So the ByteBuffer inside this DataInput implementation already contains a serialized version of ColumnFormatValue so that is transferred directly as the buffer of ColumnFormatValue avoiding any copy -- see ColumnFormatValue.fromData. Hence at this point the ColumnFormatValue's ByteBuffer contains the serialization header.

Similar changes at GemFire P2P layer were able to avoid 2-3 copies of data in reads/writes. However the same still needs to be done for Compactor and GII -- https://jira.snappydata.io/browse/SNAP-1483